### PR TITLE
Fix device_id in profiling

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/pull/4694#issuecomment-1098495097.
@khuck Can you please have a look if this fixes the issue for you?

The important change is the correct shift in `device_id_root()` to make sure that this impacts `type` and not `device_id` in `ExecutionSpaceIdentifier`. Previously, we were not recording a `device_id` at all.

### CODE REVIEW

**Review by crtrott — APPROVED** [review]

**Comment by crtrott** [comment]
You may need to instantiate the classes now to actually be able to link against the function?

File: core/src/Kokkos_OpenMPTarget.hpp [hunk 1]
```diff
@@ -141,6 +141,8 @@ template <>
 struct DeviceTypeTraits<::Kokkos::Experimental::OpenMPTarget> {
   static constexpr DeviceType id =
       ::Kokkos::Profiling::Experimental::DeviceType::OpenMPTarget;
+  // FIXME_OPENMPTARGET
+  static int device_id(const Kokkos::Experimental::OpenMPTarget&) { return 0; }
```
[the remark this answers is not in the document]
Comment (rgayatri23) [reply core/src/Kokkos_OpenMPTarget.hpp:145]: Yes thats correct. `omp_get_default_device` gives us the default target device.

**Comment by masterleinad** [comment]
> You may need to instantiate the classes now to actually be able to link against the function?

Not quite yet what the problem is...

**Comment by khuck** [comment]
> Fixes [#4694 (comment)](https://github.com/kokkos/kokkos/pull/4694#issuecomment-1098495097). @khuck Can you please have a look of this fixes the issue for you?
> 
> The important change is the correct shift in `device_id_root()` to make sure that this impacts `type` and not `device_id` in `ExecutionSpaceIdentifier`. Previously, we were not recording a `device_id` at all.

Yeah, once I get some free time after the ECP meeting, I'll give it a shot.

**Comment by masterleinad** [comment]
CI is only failing because of the changed keys for the `CUDA` containers.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Only `CUDA-10.1-Clang-Tidy` is failing (timeout), see https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/8680/pipeline/53, but `KokkosCore_UnitTest_KokkosP` (which is the one relevant for this pull request) is passing for that build as well.

**Comment by khuck** [comment]
@masterleinad yes, this appears to fix the issue.  Thanks!

File: core/src/Cuda/Kokkos_Cuda_Instance.cpp [hunk 2]
```diff
@@ -1018,6 +1018,15 @@ void CudaSpaceInitializer::print_configuration(std::ostream &msg,
 }
 
 }  // namespace Impl
+
+#ifdef KOKKOS_ENABLE_CXX14
```
Comment (Rombur) [inline core/src/Cuda/Kokkos_Cuda_Instance.cpp:1022]: Why do you need this?

File: core/src/Kokkos_OpenMP.hpp [hunk 3]
```diff
@@ -179,6 +179,7 @@ namespace Experimental {
 template <>
 struct DeviceTypeTraits<OpenMP> {
   static constexpr DeviceType id = DeviceType::OpenMP;
+  static int device_id(const OpenMP&) { return 0; }
```
Comment (Rombur) [inline core/src/Kokkos_OpenMP.hpp:182]: Shouldn't this be `omp_get_default_device()` like for OpenMPTarget?

Comment (Rombur) [reply core/src/Kokkos_OpenMP.hpp:182]: I did a quick check in the standard and I couldn't find in the standard that the CPU will always have `device_id = 0`, that's why I think it's safer to use `omp_get_default_device()`

Comment (Rombur) [reply core/src/Kokkos_OpenMP.hpp:182]: > What I was trying to say is that the device reported by omp_get_default_device is only for target regions 

What do you mean? That function does not work inside a target region. `When called from within a target region the effect of this routine is unspecified.`

Comment (masterleinad) [reply core/src/Kokkos_OpenMP.hpp:182]: Yes, but AFAICT a `device` is only used for `target` constructs, see https://splichal.eu/scripts/sphinx/libgomp/_build/html/openmp-environment-variables/ompdefaultdevice.html#omp-default-device. You use it to define a target region, not inside one.

Comment (rgayatri23) [reply core/src/Kokkos_OpenMP.hpp:182]: Could not find any documentation on the behavior of `omp_get_default_device` when a target device is not present.

**Review by rgayatri23 — APPROVED** [review]

**Pull request merged**